glib/src/closure.rs

```diff
@@ -27,17 +30,131 @@ wrapper! {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub struct RustClosure(Closure);
+
+impl RustClosure {
+    /// Creates a new closure around a Rust closure.
+    ///
+    /// See [`glib::closure!`] for a way to create a closure with concrete types.
+    ///
+    /// # Panics
+    ///
+    /// Invoking the closure with wrong argument types or returning the wrong return value type
+    /// will panic.
+    pub fn new<F: Fn(&[Value]) -> Option<Value> + Send + Sync + 'static>(callback: F) -> Self {
+        Self(Closure::new(callback))
+    }
+
+    /// Creates a new closure around a Rust closure.
+    ///
+    /// See [`glib::closure_local!`] for a way to create a closure with concrete types.
+    ///
+    /// # Panics
+    ///
+    /// Invoking the closure with wrong argument types or returning the wrong return value type
+    /// will panic.
+    ///
+    /// Invoking the closure from a different thread than this one will panic.
+    pub fn new_local<F: Fn(&[Value]) -> Option<Value> + 'static>(callback: F) -> Self {
+        Self(Closure::new_local(callback))
+    }
+
+    /// Invokes the closure with the given arguments.
+    ///
+    /// For invalidated closures this returns the "default" value of the return type. For nullable
+    /// types this is `None`, which means that e.g. requesting `R = String` will panic will `R =
+    /// Option<String>` will return `None`.
+    ///
+    /// # Panics
+    ///
+    /// The argument types and return value type must match the ones expected by the closure or
+    /// otherwise this function panics.
+    pub fn invoke<R: TryFromClosureReturnValue>(&self, values: &[&dyn ToValue]) -> R {
+        let values = values
+            .iter()
+            .copied()
+            .map(ToValue::to_value)
+            .collect::<smallvec::SmallVec<[_; 10]>>();
+
+        R::try_from_closure_return_value(self.invoke_with_values(R::static_type(), &values))
+            .expect("Invalid return value")
+    }
+
+    /// Invokes the closure with the given arguments.
+    ///
+    /// For invalidated closures this returns the "default" value of the return type.
+    ///
+    /// # Panics
+    ///
+    /// The argument types and return value type must match the ones expected by the closure or
+    /// otherwise this function panics.
+    pub fn invoke_with_values(&self, return_type: Type, values: &[Value]) -> Option<Value> {
+        unsafe { self.0.invoke_with_values(return_type, values) }
+    }
+
+    /// Invalidates the closure.
+    ///
+    /// Invoking an invalidated closure has no effect.
+    pub fn invalidate(&self) {
+        self.0.invalidate();
+    }
+}
+
+impl From<RustClosure> for Closure {
+    fn from(c: RustClosure) -> Self {
+        c.0
+    }
+}
+
+impl AsRef<Closure> for RustClosure {
+    fn as_ref(&self) -> &Closure {
+        &self.0
+    }
+}
+
+impl AsRef<Closure> for Closure {
```

Comment:
This should behind a `#[doc(hidden)]` maybe?

Reply:
I don't like that we're hiding trait impls in the docs in general.

Reply:
I don't think it matter much here...